### ISSUE TITLE
[FR] SDK 支持 forward_backward_custom（自定义 Loss Function）

### ISSUE THREAD

**Issue by Etoilelune** [nex-agi/weaver#42]
## 背景

来自 feedback: china-qijizhifeng/nex-taas-feedback#90

参考 [Tinker 的 forward_backward_custom](https://tinker-docs.thinkingmachines.ai/losses#flexible-loss-functions-forward_backward_custom) 设计，用户在客户端定义 loss function，SDK 编排 forward/backward 拆分执行。

## 使用场景

```python
from weaver import TrainingClient
import torch

training = TrainingClient(...)

# 内置 loss（现有，不变）
training.forward_backward(data, loss_fn='cross_entropy')

# 自定义 loss（新增）
def my_custom_loss(logprobs, datum):
    # 用户自定义 loss 计算逻辑，在客户端执行
    advantages = datum.loss_fn_inputs['advantages']
    loss = -(logprobs * advantages).sum()
    return loss

result = training.forward_backward_custom(data, loss_fn=my_custom_loss)
```

## 预期行为

- 新增 `TrainingClient.forward_backward_custom(datum, loss_fn)` 方法
- 内部编排：
  1. 调用 Server forward API → 获取 logprobs（`ForwardOutput`）
  2. 在客户端执行用户的 `loss_fn(logprobs, datum)` → 得到 loss tensor
  3. 调用 Server backward API → 将 loss 发回 trainer 执行 backward pass
- loss_fn 签名约束：接收 logprobs（tensor）和 datum，返回标量 loss
- 提供 `forward_backward_custom_async` 异步版本
- 文档：说明 custom loss 会比内置 loss 多一次网络往返，适用于实验性 loss 设计

## Tinker 参考

Tinker 支持两种方式：
- `forward_backward(loss_fn='cross_entropy')` — 内置 loss，一步完成
- `forward_backward_custom(loss_fn=callable)` — 自定义 loss，拆分执行，多一次 forward pass 开销

内置 loss 包括：cross_entropy、importance_sampling、ppo、cispo、dro

## 关键验证方法

- 用户传入自定义 loss callable → SDK 正确编排 forward/backward → 返回与手动计算一致的 loss 值和梯度
- 现有 `forward_backward` 不受影响

## 关联

- Parent: china-qijizhifeng/nex-taas-feedback#90
- Server 侧: china-qijizhifeng/weaver-server#182
- Trainer 侧: china-qijizhifeng/weaver-trainer#105
